File: 프로그래머스/3/43238. 입국심사/입국심사.java
import java.util.*;

class Solution {
    static long people = 0;
    static int[] testTimes;
    
    public long solution(int n, int[] times) {
        testTimes = times;
        people = n;
        
        Arrays.sort(testTimes);
        
        long min = 1;
        long max = (long) testTimes[testTimes.length - 1] * n; 
        
        long answer = BSearch(min, max);
        return answer;
    }
    
    public static long BSearch(long min, long max) {
        long answer = max;
        while(min <= max) {
            long mid = (min + max) / 2;
            long possiblePeople = 0;
            
            for(int time : testTimes) {
                possiblePeople += mid / time;
            }
            
            if(possiblePeople >= people) {
                answer = mid;
                max = mid - 1;
            } else {
                min = mid + 1;
            }
        }
        return answer;
    }
}

File: 프로그래머스/3/43238. 입국심사/입국심사.js
function solution(n, times) {
    var answer = 0;
    
    var min = 0;
    var max = n * Math.max(...times);
    
    while (min <= max) {
        let mid = Math.floor((min + max) / 2);
        let people = 0;
        
        for (let time of times) {  
            people += Math.floor(mid / time);
        }
        
        if (people >= n) { 
            answer = mid;
            max = mid - 1;
        } else {  
            min = mid + 1;
        }
    }
    
    return answer;
}
